**TRABALHO 10 - ASSOCIAÇÃO**
**Grupo:** Luis Felipe, Marcella Resende, Sofia Castilho

In [ ]:
%pip install mlxtend --upgrade

In [ ]:
import pandas as pd
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori, association_rules
from mlxtend.frequent_patterns import fpgrowth
import itertools

In [ ]:
import warnings
warnings.filterwarnings('ignore')

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
url = 'https://raw.githubusercontent.com/marcellaresende/AssociationRules/main/BreadBasket_DMS.csv'

df = pd.read_csv(url)

In [ ]:
transactions = df[df['Item'] != 'NONE'].groupby('Transaction')['Item'].apply(list)


te = TransactionEncoder()
data_encoded = te.fit_transform(transactions)
df_encoded = pd.DataFrame(data_encoded, columns=te.columns_)

In [ ]:
df_encoded.head()

Adjustment  Afternoon with the baker  Alfajores  Argentina Night  Art Tray  \
0       False                     False      False            False     False   
1       False                     False      False            False     False   
2       False                     False      False            False     False   
3       False                     False      False            False     False   
4       False                     False      False            False     False   

   Bacon  Baguette  Bakewell  Bare Popcorn  Basket  ...  The BART  The Nomad  \
0  False     False     False         False   False  ...     False      False   
1  False     False     False         False   False  ...     False      False   
2  False     False     False         False   False  ...     False      False   
3  False     False     False         False   False  ...     False      False   
4  False     False     False         False   False  ...     False      False   

   Tiffin  Toast  Truffles  Tshirt  Valentine's card  Vegan Feast  \
0   False  False     False   False             False        False   
1   False  False     False   False             False        False   
2   False  False     False   False             False        False   
3   False  False     False   False             False        False   
4   False  False     False   False             False        False   

   Vegan mincepie  Victorian Sponge  
0           False             False  
1           False             False  
2           False             False  
3           False             False  
4           False             False  

[5 rows x 94 columns]

In [ ]:
min_support = 0.01
frequent_itemsets_apriori = apriori(df_encoded, min_support=min_support, use_colnames=True, verbose = 1)
frequent_itemsets_apriori.sort_values(by = "support", ascending = False)

Processing 12 combinations | Sampling itemset size 4


support               itemsets
6   0.478394               (Coffee)
2   0.327205                (Bread)
26  0.142631                  (Tea)
4   0.103856                 (Cake)
34  0.090016        (Coffee, Bread)
..       ...                    ...
11  0.010565    (Hearty & Seasonal)
20  0.010460                (Salad)
30  0.010354     (Bread, Alfajores)
58  0.010037  (Coffee, Cake, Bread)
60  0.010037    (Coffee, Cake, Tea)

[61 rows x 2 columns]

In [ ]:
rules = association_rules(frequent_itemsets_apriori, metric= "confidence", min_threshold = 0.5)

rules = rules.sort_values(['confidence', 'lift'], ascending =[False, False])

rules

antecedents consequents  antecedent support  consequent support  \
10           (Toast)    (Coffee)            0.033597            0.478394   
9   (Spanish Brunch)    (Coffee)            0.018172            0.478394   
5        (Medialuna)    (Coffee)            0.061807            0.478394   
6           (Pastry)    (Coffee)            0.086107            0.478394   
0        (Alfajores)    (Coffee)            0.036344            0.478394   
4            (Juice)    (Coffee)            0.038563            0.478394   
7         (Sandwich)    (Coffee)            0.071844            0.478394   
1             (Cake)    (Coffee)            0.103856            0.478394   
8            (Scone)    (Coffee)            0.034548            0.478394   
2          (Cookies)    (Coffee)            0.054411            0.478394   
3    (Hot chocolate)    (Coffee)            0.058320            0.478394   

     support  confidence      lift  leverage  conviction  zhangs_metric  
10  0.023666    0.704403  1.472431  0.007593    1.764582       0.332006  
9   0.010882    0.598837  1.251766  0.002189    1.300235       0.204851  
5   0.035182    0.569231  1.189878  0.005614    1.210871       0.170091  
6   0.047544    0.552147  1.154168  0.006351    1.164682       0.146161  
0   0.019651    0.540698  1.130235  0.002264    1.135648       0.119574  
4   0.020602    0.534247  1.116750  0.002154    1.119919       0.108738  
7   0.038246    0.532353  1.112792  0.003877    1.115384       0.109205  
1   0.054728    0.526958  1.101515  0.005044    1.102664       0.102840  
8   0.018067    0.522936  1.093107  0.001539    1.093366       0.088224  
2   0.028209    0.518447  1.083723  0.002179    1.083174       0.081700  
3   0.029583    0.507246  1.060311  0.001683    1.058553       0.060403

**b) Apresente a interpretação (detalhada) de cada uma das regras geradas:**

**Regra Toast -> Coffee:**  Há uma associação positiva entre pedidos de Toast e pedidos Coffee, com uma confiança de 70.44%. O valor do lift de 1.4724 sugere que a ocorrência conjunta de Toast e Coffee é 1.4724 vezes mais provável do que o esperado se as duas escolhas fossem separadas.

**Regra Spanish Brunch -> Coffee:** Há uma associação positiva entre pedidos de Spanish Brunch e pedidos de Coffee, com uma confiança de 59.88%. O valor do lift de 1.2518 sugere que a ocorrência conjunta do Spanish Brunch e Coffee é 1.2518 vezes mais provável do que o esperado se as duas escolhas fossem feitas separadamente.

**Regra Medialuna -> Coffee:** Há uma associação positiva entre pedidos de medialuna e Coffee, com uma confiança de 56.92%. O valor do lift de 1.1899 sugere que a ocorrência conjunta de medialuna e Coffee é 1.1899 vezes mais provável do que o esperado se as duas escolhas fossem independentes.

**Regra Pastry -> Coffee:** Há uma associação positiva entre pedidos de Pastry e Coffe, com uma confiança de 55.21%. O valor do lift de 1.1542 sugere que a ocorrência conjunta destes é 1.1542 vezes mais provável do que o esperado se as duas escolhas fossem feitas separadamente.

**Regra Alfajores -> Coffee:** Há uma associação positiva entre pedidos de alfajores e Coffeecom uma confiança de 54.07%. O valor do lift de 1.1302 sugere que a ocorrência conjunta de alfajores e Coffee é 1.1302 vezes mais provável do que o esperado se as duas escolhas fossem separadas.

**Regra Juice -> Coffe:**  Há uma associação positiva entre pedidos de Juice e Coffe, com uma confiança de 53.42%. O valor do lift de 1.1167 sugere que a ocorrência conjunta destes é 1.1167 vezes mais provável do que o esperado se as duas escolhas fossem independentes.

**Regra Sandwich -> Coffe:** Há uma associação positiva entre pedidos de Sandwich e Coffe, com uma confiança de 53.24%. O valor do lift de 1.1128 sugere que a ocorrência conjunta de Sandwich e Coffe é 1.1128 vezes mais provável do que o esperado se as duas escolhas fossem feitas separadas.

**Regra Cake -> Coffee:** Há uma associação positiva entre pedidos de Cake e Coffee com uma confiança de 52.70%. O valor do lift de 1.1015 sugere que a ocorrência conjunta de Cake e Coffee é 1.1015 vezes mais provável do que o esperado se as duas escolhas fossem separadas.

**Regra Scone -> Coffee:** Há uma associação positiva entre pedidos de Scone e Coffee, com uma confiança de 52.29%. O valor do lift de 1.0931 sugere que a ocorrência conjunta destes é 1.0931 vezes mais provável do que o esperado se as duas escolhas fossem independentes.

**Regra Cookies -> Coffee:** Há uma associação positiva entre pedidos de Cookies e Coffee, com uma confiança de 51.84%. O valor do lift de 1.0837 sugere que a ocorrência conjunta de Cookies -> Coffee é 1.0837 vezes mais provável do que o esperado se as duas escolhas fossem separadas.

**Regra Hot chocolate -> Coffee:** Há uma associação positiva entre pedidos de Hot chocolate e Coffee, com uma confiança de 50.72%. O valor do lift de 1.0603 sugere que a ocorrência conjunta destes é 1.0603 vezes mais provável do que o esperado se as duas escolhas fossem independentes.

Em todas as regras o valor leverage é positivo, indicando uma associação positiva significativa entre as duas escolhas.




In [ ]:
min_support = 0.02
frequent_itemsets_fpgrowth = fpgrowth(df_encoded, min_support=min_support, use_colnames=True,  verbose=1)
frequent_itemsets_fpgrowth.sort_values(by = "support", ascending = False)


19 itemset(s) from tree conditioned on items ()
1 itemset(s) from tree conditioned on items (Bread)
0 itemset(s) from tree conditioned on items (Scandinavian)
1 itemset(s) from tree conditioned on items (Hot chocolate)
1 itemset(s) from tree conditioned on items (Cookies)
0 itemset(s) from tree conditioned on items (Muffin)
0 itemset(s) from tree conditioned on items (Coffee)
2 itemset(s) from tree conditioned on items (Pastry)
0 itemset(s) from tree conditioned on items (Pastry, Coffee)
0 itemset(s) from tree conditioned on items (Pastry, Bread)
1 itemset(s) from tree conditioned on items (Medialuna)
2 itemset(s) from tree conditioned on items (Tea)
0 itemset(s) from tree conditioned on items (Tea, Coffee)
0 itemset(s) from tree conditioned on items (Tea, Bread)
0 itemset(s) from tree conditioned on items (Farm House)
1 itemset(s) from tree conditioned on items (Juice)
0 itemset(s) from tree conditioned on items (Soup)
3 itemset(s) from tree conditioned on items (Cake)
0 itemset(s) fr

support                 itemsets
5   0.478394                 (Coffee)
0   0.327205                  (Bread)
8   0.142631                    (Tea)
12  0.103856                   (Cake)
19  0.090016          (Bread, Coffee)
6   0.086107                 (Pastry)
13  0.071844               (Sandwich)
7   0.061807              (Medialuna)
2   0.058320          (Hot chocolate)
28  0.054728           (Cake, Coffee)
3   0.054411                (Cookies)
25  0.049868            (Tea, Coffee)
22  0.047544         (Pastry, Coffee)
15  0.040042                (Brownie)
9   0.039197             (Farm House)
10  0.038563                  (Juice)
4   0.038457                 (Muffin)
31  0.038246       (Sandwich, Coffee)
14  0.036344              (Alfajores)
24  0.035182      (Medialuna, Coffee)
18  0.034548                  (Scone)
11  0.034443                   (Soup)
17  0.033597                  (Toast)
20  0.029583  (Hot chocolate, Coffee)
23  0.029160          (Pastry, Bread)
1   0.029054           (Scandinavian)
21  0.028209        (Cookies, Coffee)
26  0.028104             (Tea, Bread)
29  0.023772              (Cake, Tea)
32  0.023666          (Toast, Coffee)
30  0.023349            (Cake, Bread)
27  0.020602          (Juice, Coffee)
16  0.020285               (Truffles)

In [ ]:
rules = association_rules((frequent_itemsets_fpgrowth), metric="confidence", min_threshold=0.5)

rules = rules.sort_values(['confidence', 'lift'], ascending =[False, False])

rules

antecedents consequents  antecedent support  consequent support  \
7          (Toast)    (Coffee)            0.033597            0.478394   
3      (Medialuna)    (Coffee)            0.061807            0.478394   
2         (Pastry)    (Coffee)            0.086107            0.478394   
4          (Juice)    (Coffee)            0.038563            0.478394   
6       (Sandwich)    (Coffee)            0.071844            0.478394   
5           (Cake)    (Coffee)            0.103856            0.478394   
1        (Cookies)    (Coffee)            0.054411            0.478394   
0  (Hot chocolate)    (Coffee)            0.058320            0.478394   

    support  confidence      lift  leverage  conviction  zhangs_metric  
7  0.023666    0.704403  1.472431  0.007593    1.764582       0.332006  
3  0.035182    0.569231  1.189878  0.005614    1.210871       0.170091  
2  0.047544    0.552147  1.154168  0.006351    1.164682       0.146161  
4  0.020602    0.534247  1.116750  0.002154    1.119919       0.108738  
6  0.038246    0.532353  1.112792  0.003877    1.115384       0.109205  
5  0.054728    0.526958  1.101515  0.005044    1.102664       0.102840  
1  0.028209    0.518447  1.083723  0.002179    1.083174       0.081700  
0  0.029583    0.507246  1.060311  0.001683    1.058553       0.060403

**b) Apresente a interpretação (detalhada) de cada uma das regras geradas:**

**Regra Toast -> Coffee:** Há uma associação positiva entre pedidos de torrada (toast) e pedidos de café. A confiança de 0.7044 indica que, quando um cliente pede torrada, há uma chance de 70.44% de que eles também peçam café. O valor do lift de 1.4724 sugere que a ocorrência conjunta de torrada e café é 1.4724 vezes mais provável do que o esperado se as duas escolhas fossem independentes.

**Regra medialuna -> Coffee:** Há uma associação positiva entre pedidos de medialuna e Coffee, com uma confiança de 56.92%. O valor do lift de 1.1899 sugere que a ocorrência conjunta destes é 1.1899 vezes mais provável do que o esperado se as duas escolhas fossem feitas separadas.

**Regra Pastry -> Coffee:** Há uma associação positiva entre pedidos de Pastry e Coffee, com uma confiança de 55.21%. O valor do lift de 1.1542 sugere que a ocorrência conjunta de Pastry e Coffee é 1.1542 vezes mais provável do que o esperado se as duas escolhas fossem separadas.

**Regra Juice -> Coffee:** Há uma associação positiva entre pedidos de Juice e Coffee com uma confiança de 53.42%. O valor do lift de 1.1167 sugere que a ocorrência conjunta de Juice e Coffee é 1.1167 vezes mais provável do que o esperado se as duas escolhas fossem independentes.

**Regra Sandwich -> Coffee:** Há uma associação positiva entre pedidos de Sandwich e Coffee, com uma confiança de 53.24%. O valor do lift de 1.1128 sugere que a ocorrência conjunta destes é 1.1128 vezes mais provável do que o esperado se as duas escolhas fossem separadas.

**Regra Cake -> Coffee:** Há uma associação positiva entre pedidos de Cake e Coffee, com confiança de 52.69%. O valor do lift de 1.1015 sugere que a ocorrência conjunta de Cake e Coffee é 1.1015 vezes mais provável do que o esperado se as duas escolhas fossem feitas separadamente.

 **Regra Cookies -> Coffee:** Há uma associação positiva entre pedidos de Cookies e Coffee com confiança de 51.84%. O valor do lift de 1.0837 sugere que a ocorrência conjunta de Cookies e Coffee é 1.0837 vezes mais provável do que o esperado se as duas escolhas fossem independentes.

 **Regra Hot Chocolate -> Coffee:** Há uma associação positiva entre pedidos de Hot Chocolate e Coffee, com uma confiança de 50.72%. O valor do lift de 1.0603 sugere que a ocorrência conjunta destes é 1.0603 vezes mais provável do que o esperado se as duas escolhas fossem separados.

 Em todas as regras o valor leverage é positivo, indicando uma associação positiva significativa entre as duas escolhas.

**Passo 5**

**a) Quais as diferenças entre as regras geradas pelos dois algoritmos?**
A diferença nesse caso ficou quanto a abordagem técnica dos algoritmos, uma vez que as regras geradas foram muito semelhantes.

**b) Para ambos os algoritmos, realize testes com 4 diferentes valores de suporte mínimo e confiança mínima. O que acontece? Por que?**

Em resumo, quando alteramos os valores de suporte e confiança isso impacta diretamente na quantidade e qualidade das regras geradas pelos algoritmos.

In [ ]:
min_support = [0.01, 0.02, 0.03, 0.04]
min_confidence = [0.5, 0.6, 0.7, 0.8]

In [ ]:
for support, confidence in itertools.product(min_support, min_confidence):
    frequent_itemsets_apriori = apriori(df_encoded, min_support=support, use_colnames=True, verbose=0)
    rules_apriori = association_rules(frequent_itemsets_apriori, metric="confidence", min_threshold=confidence)
    print(f"Apriori - Support: {support}, Confidence: {confidence}")
    print(rules_apriori)
    print("-----------------------------------------")


Apriori - Support: 0.01, Confidence: 0.5
         antecedents consequents  antecedent support  consequent support  \
0        (Alfajores)    (Coffee)            0.036344            0.478394   
1             (Cake)    (Coffee)            0.103856            0.478394   
2          (Cookies)    (Coffee)            0.054411            0.478394   
3    (Hot chocolate)    (Coffee)            0.058320            0.478394   
4            (Juice)    (Coffee)            0.038563            0.478394   
5        (Medialuna)    (Coffee)            0.061807            0.478394   
6           (Pastry)    (Coffee)            0.086107            0.478394   
7         (Sandwich)    (Coffee)            0.071844            0.478394   
8            (Scone)    (Coffee)            0.034548            0.478394   
9   (Spanish Brunch)    (Coffee)            0.018172            0.478394   
10           (Toast)    (Coffee)            0.033597            0.478394   

     support  confidence      lift  leverage  

In [ ]:
for support, confidence in itertools.product(min_support, min_confidence):
    frequent_itemsets_fpgrowth = fpgrowth(df_encoded, min_support=support, use_colnames=True, verbose=0)
    rules_fpgrowth = association_rules(frequent_itemsets_fpgrowth, metric="confidence", min_threshold=confidence)
    print(f"FP-Growth - Support: {support}, Confidence: {confidence}")
    print(rules_fpgrowth)
    print("-----------------------------------------")

FP-Growth - Support: 0.01, Confidence: 0.5
         antecedents consequents  antecedent support  consequent support  \
0    (Hot chocolate)    (Coffee)            0.058320            0.478394   
1          (Cookies)    (Coffee)            0.054411            0.478394   
2           (Pastry)    (Coffee)            0.086107            0.478394   
3        (Medialuna)    (Coffee)            0.061807            0.478394   
4            (Juice)    (Coffee)            0.038563            0.478394   
5             (Cake)    (Coffee)            0.103856            0.478394   
6         (Sandwich)    (Coffee)            0.071844            0.478394   
7        (Alfajores)    (Coffee)            0.036344            0.478394   
8            (Toast)    (Coffee)            0.033597            0.478394   
9            (Scone)    (Coffee)            0.034548            0.478394   
10  (Spanish Brunch)    (Coffee)            0.018172            0.478394   

     support  confidence      lift  leverage